Add support for Apache Drill

cgivre opened the pull request:
This is the first attempt at support for Apache Drill in Superset.  This uses John Omernik's SQLAlchemy dialect (https://github.com/JohnOmernik/sqlalchemy-drill).

Drill has a unique aspect in its namespace which proved problematic.  Drill uses the concepts of:

- Storage Plugin
- Workspace
- Table (or file)

When connecting to Drill to a non-file based system, the behavior is pretty much as a traditional database.  However, when Drill is querying a file based system, you can have additional `.` in the namespace that are not namespace related.   For example:  
```
dfs.test.`data.json`
```
In this case, the storage plugin is `dfs`, the workspace which is optional is `test`, and the table, which in this case is a file, is `data.json`.  In this case however, the last period is not part of the namespace and was consistently misinterpreted by Superset.  

The solution I found, which seemed like complete hackery, was to replace the last period with `@@@` in the dropdown in SQLlab.  Then I modified `view/core.py` to replace the `@@@` with a period in the various functions that generate tables.   

**UPDATE:  All Hackery Removed**

mistercrunch commented on superset/views/core.py:
```diff
@@ -1560,8 +1560,16 @@ def tables(self, db_id, schema, substr, force_refresh='false'):
+        # Make modifications for Drill
+        # This hack (and it's a total hack) replaces the dot from the file extension
+        # with '@@@' to prevent Superset from misinterpreting the schema
+        if "drill+sadrill" in database.sqlalchemy_uri_decrypted:
```
To avoid having `if database.engine == 'some_engine'` all over the codebase, we make sure to keep all of the database engine specific code in `db_engine_specs.py`. We may need to add some new hook in `db_engine_specs.py` to support your use case, though hopefully not a hack...

cgivre replied:
@mistercrunch 
Perhaps hack was too strong a word, but I couldn't find an appropriate hook. 
The drop down is now populated as follows:
<option value="filename@@@csv">filename.csv</option>

Do you have any suggestions as to either how to write a hook to properly encode/decode the URL or otherwise get the arguments passed correctly?

mistercrunch replied:
I mean you have to introduce a new hook to `db_engine_spec.BaseEngineSpec`, so that all the oddness of this hack (or whatever you'll call it) is limited to the `DrillEngineSpec` class. Something like `db_engine_spec.alter_table_name(table_name)`

mistercrunch replied:
The goal is to not have any database-engine-specific logic outside of `db_engines_spec.py`, so that implementing a new engine is limited to creating a new class in that module, instead of altering methods scattered all around the code base.

cgivre replied:
Fixed/

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -590,6 +590,58 @@ def get_table_names(cls, inspector, schema):
+    def extra_table_metadata(cls, database, table_name, schema_name):
+        """Returns engine-specific table metadata"""
+
+        return {}
```
I doin't think you need to override as the base class does the same

cgivre replied:
Fixed

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -590,6 +590,58 @@ def get_table_names(cls, inspector, schema):
+                    cols=None):
+        table_name = table_name.replace('@@@', '.')
+
+        fields = '*'
```
This method looks largely copy-pastaed from the base class, why not just calling `super` with the altered `table_name`?

cgivre replied:
Fixed

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -590,6 +590,58 @@ def get_table_names(cls, inspector, schema):
+    def select_star(cls, my_db, table_name, engine, schema=None, limit=100,
+                    show_cols=False, indent=True, latest_partition=True,
+                    cols=None):
+        table_name = table_name.replace('@@@', '.')
```
A comment would be nice here, and not hard coding would be nice too. You could use a constant in module scope or better, import the constant from the related dialect package. Something like `from sqlalchemy_drill import DRILL_SCHEMA_TABLE_DELIMITER` or whatever it would be declared as on the other side.

Wherever that constant is declared or used, it could be good to reference documentation that explains the hack and why it's necessary.

cgivre replied:
Fixed.

cgivre commented:
There is one small issue with this PR that I would really like to resolve but I'm a little stuck on. 
I have Superset connected to Drill and Drill connected to a MySQL database.  Superset is correctly getting the tables, and schemata!
(Yay!)
When you try to access a file-based table via Drill, there’s one small problem.
As I mentioned in the PR, the issue is that filenames have dots in them, so when the schema was being passed to Superset you might have something like this
`dfs.test.somefile.csv`
The problem being that when Superset parses that, it parses that fourth period as a part of the namespace rather than a file extension
So… the workaround I found was, in the dropdown in sqllab, replace the value (not the label) with `@@@` so that it gets parsed correctly.
This works as shown below:

![screen shot 2019-01-21 at 16 46 51](https://user-images.githubusercontent.com/5513150/51756888-099a5700-2090-11e9-9ece-4c54cc397535.png)

Here's where I'm stuck. 

In all the other functions where I could find this, I added a hook to clean up the file name, specifically, to look for `@@@` and replace it with a `.`.
I can’t seem to find the code that generates the heading for the table preview and the heading for the column listing.   Once I can figure this out, I think Drill will pretty much work perfectly with Superset!!  Any suggestions would be greatly appreciated.  @mistercrunch @kristw 

![screen shot 2019-01-21 at 16 48 10](https://user-images.githubusercontent.com/5513150/51757144-a4933100-2090-11e9-8c0f-567080542a7e.png)

Thanks!!

cgivre commented:
@mistercrunch @kristw 
Hello all, 
I managed to get everything working properly!   I think this is ready for review.

cgivre commented:
@mistercrunch  All unit tests passed.  Superset + Drill is a really awesome combination.  Do you want me to squash all commits?

mistercrunch commented:
I don't think it's ok to have references to `@@@` scattered across the code base. It really should be limited to the `db_engine_spec` entry for drill. Looks like there are few entries on the frontend and one on the backend outside of `db_engine_spec` to clean up.

JohnOmernik commented:
I am watching this PR very closely. I think this is a great addition to Superset and look forward to seeing it merged.

cgivre commented:
Hi @mistercrunch ,
I'm a little stuck and wondering if perhaps you could point me in the right direction.  
I've removed all the static references to `@@@` in the code base, instead calling the function in `db_engine_specs` with the exception of two that I can't seem to figure out. 

They are: superset/assets/src/SqlLab/components/SouthPane.jsx and superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx.

I can't seem to figure out how these functions are populated or called from the Python/Flask side. 

```
onTableChange(tableName, schemaName) {
	    this.props.actions.addTable(this.props.queryEditor, tableName, schemaName);
	    // Mod for Apache Drill
	    const cleanTableName = tableName.replace('@@@', '.');
	    this.props.actions.addTable(this.props.queryEditor, cleanTableName, schemaName);
	  }
```

If you could point me in the right direction there, and show me where in the python code these functions are getting the data, I can wrap this up. 
Thanks in advance.

mistercrunch commented:
@cgivre I'm happy to help push that through and providing pointers. So here's the view:
https://github.com/apache/incubator-superset/blob/master/superset/views/core.py#L2303 and more specifically around here?
https://github.com/apache/incubator-superset/blob/master/superset/views/core.py#L2349

I didn't dig super deep as I don't have time to dig right this moment, but wanted to give some pointers. 
 Let me know if the pointers aren't helpful, I'll take the time to dig in deeper.

cgivre commented:
@mistercrunch Thanks for your response.   The PR already had a hook there to clean up the table name (see below)

https://github.com/apache/incubator-superset/blob/a05fc76633a93ccc6a33ac0badd9f3969cca35bf/superset/views/core.py#L2311

I've went through most of the flles looking for anything like `table_name` or `tmp_table_name` and run it through the clean up hook.  Any help is greatly appreciated.

cgivre commented:
@mistercrunch @kristw 
I'm wondering if we could mark this as a **Work In Progress** and integrate this PR.  The latest version has two places in the code with the `@@@` hard coded.  

https://github.com/apache/incubator-superset/blob/184c6c625090da91dedff151f12ed560d83f2385/superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx#L63-64

and

https://github.com/apache/incubator-superset/blob/184c6c625090da91dedff151f12ed560d83f2385/superset/assets/src/SqlLab/components/SouthPane.jsx#L103-107

There are some things which I'd like to contribute but really can't until this is committed.  I'll keep working on it and hopefully get some more help as more people start using Drill+Superset. 

@JohnOmernik

kristw commented:
@cgivre I am not familiar with the backend code so I cannot fully review the code.

However, regarding your request to merge as **Work in Progress**, Superset is used in production environment for many organizations, making `master` stability very important. Therefore, I will have a hard time merging incompleted code into master.  PR are usually labelled **WIP (Work in Progress)** when the work is not completed so reviewers will not even review the PR.

However, if you feel that there are parts of this PR that are solid already and ready to be merged, please keep only that part and get rid of everything else. You can also help the reviewers understand what are the remaining issues that make you say this is still WIP. 

If you can clean out the questionable parts, after receiving approval from @mistercrunch or other backend experts, we can merge.

cgivre commented:
Hi @kristw 
I do not believe that this PR will introduce any instability or is not production ready.  As I indicated, it passes all the unit tests and works perfectly when I've used it myself.  

The remaining issue is that @mistercrunch prefers to have code that is specific to one database in the `db_engine_specs.py` file.  (Which makes perfect sense and something I completely understand)   However, there are two places in this PR that I could not get to work without putting database specific code into another file.   If someone can show me how I can fix that, I will gladly do so, but I'm really stuck at this point. 

As I said, I do feel this PR is solid and ready to be merged.  I'm willing to keep working on it to figure out how to move this code to the `db_engine_specs.py` file.

cgivre commented:
Bump... Any update?

villebro commented:
@cgivre maybe I can help; @yciabaud recently ran into something similar with Presto+Pulsar which was addressed in #7297 . While slightly different there seem to be parallels. Can you summarize where you are stuck right now to help me get up to speed on your issue? By the looks of it this is quite close to an elegant solution but still needs that last finishing touch.

cgivre commented:
Hi @villebro 
Thanks for responding!  Basically, I added a function to the `db_engine_specs` file that replaces the file extension with `@@@`.  Then I wrote another function in the `db_engine_specs` that removes it.  

What was happening was that Superset was working great with Drill when it was querying a database, but if something had a file, the dot for the file extension would mess up the URLs.  

My approach works, however, there were two places that I couldn't figure out how to call the database specific function in `db_engine_specs` so I just added the string replacement in the code.   The two locations are:
https://github.com/apache/incubator-superset/blob/ce3f9c30a925330763b394903d1a16ce6a90a495/superset/assets/src/SqlLab/components/SouthPane.jsx#L107

and 

https://github.com/apache/incubator-superset/blob/ce3f9c30a925330763b394903d1a16ce6a90a495/superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx#L64

As you can see, these are not python files.  I couldn't figure out how to insert the Python function to clean up the table name.  

Another approach which I tried but couldn't figure out was to add a function in db_engine_specs to do all the URL transformations.  

Any assistance would be greatly appreciated, as I'd really like to get this integrated into Superset.
Thanks!!

villebro commented:
It looks like the culprit is here: 
https://github.com/apache/incubator-superset/blob/5f28027ce745dfda67e18fe7df8468184400be50/superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx#L86-L96
It seems the table selector is assuming that a table name with a period is schema.table, hence omitting everything before the period. Blame links it to a PR that's two years old: https://github.com/apache/incubator-superset/pull/1466
@mistercrunch do you have any idea or context on what's going on here?

villebro commented:
@kristw can you check out my last comment; is `SqlEditorLeftBar` deprecated, and is my interpretation correct that a table can't be chosen unless a schema is first chosen, making the `namePieces` logic obsolete?

coderfender commented:
@villebro , I was also working on this issue this weekend and figured that this is an issue on JS side.

cgivre commented:
Hi @villebro, 
You are correct in that a table can't be chosen unless a schema is first selected. 
@coderfender 
If what you're saying is correct, is there any way to fix this on the Python side?

villebro commented:
I suggest making the necessary changes (removing the `@@@` stuff, removing `namePieces` logic etc) and pushing them here. It will make it easier to review the PR, see what (if anything) breaks and also for others to try it out.

cgivre commented:
@villebro Would you mind please posting your code for those sections?

coderfender commented:
@cgivre , @villebro , You guys are rockstars ! . @cgivre , Shall I go ahead and test so that we can go ahead to raise PR and merge and probably get this in next release?

villebro commented:
Great stuff! However, let me test this first before declaring victory, there are a few bits that I'm worried might be affected that I'd like to clear first.

villebro commented:
@cgivre it seems there's still some hackery remnants (`get_table_name_dropdown`, `clean_table_name` and multiple references to it that don't do anything) and unneeded non-functional changes (added line changes) left. Please try to remove those to make this PR as compact as possible.

cgivre commented:
@villebro I think I got rid of all the hackery this time.  I also added better time grains, but should note that this will only work with Drill v 1.16 or greater.
There is one unit test which is now failing, the javascript one, which I need to sort out.

cgivre commented:
Ok, @villebro, @coderfender  Last question.  I made the changes and now the unit test for javascript is breaking.

Here is the error I'm getting:
```
FAIL spec/javascripts/components/TableSelector_spec.jsx (6.083s)
  ● TableSelector › changeTable › test 2

    expect(received).toBe(expected) // Object.is equality

    Expected: "my_table"
    Received: "main.my_table"

      200 |         label: 'my_table',
      201 |       });
    > 202 |       expect(mockedProps.onTableChange.getCall(0).args[0]).toBe('my_table');
          |                                                            ^
      203 |       expect(mockedProps.onTableChange.getCall(0).args[1]).toBe('main');
      204 |     });
      205 |   });

      at Object.toBe (spec/javascripts/components/TableSelector_spec.jsx:202:60)1
```
I'm guessing that the issue is something silly with the file below, but I'm not really sure how to fix that.

https://github.com/apache/incubator-superset/blob/81c2c896f1a2b4479ed8a4614e7268d606a66401/superset/assets/src/components/TableSelector.jsx#L173-L174

coderfender commented:
Hello @cgivre .. I think I found the place where we need to change the code. If my understanding is correct , we need to change the file TableSelector_spec.jsx like the below screenshot

![image](https://user-images.githubusercontent.com/11091419/56545999-21b63100-652e-11e9-8f8a-a30c017d5274.png)

Given my limited skills in frontend tech, I would be happy if some one can comment on the same.

Thank you

cgivre committed:
Add support for Apache Drill

cgivre commented:
Ok, this is weird.  The unit tests all pass on my machine, but when I submit the PR, I get the following errors. 

```

  1) Dashboard save "before each" hook for "should save/overwrite dashboard":
     Error: Blocked a frame with origin "http://localhost:8081" from accessing a cross-origin frame.
      at jquery (http://localhost:8081/__cypress/runner/cypress_runner.js:58622:46)
      at Object.getRemotejQueryInstance (http://localhost:8081/__cypress/runner/cypress_runner.js:58627:22)
      at Context.thenFn (http://localhost:8081/__cypress/runner/cypress_runner.js:53952:24)
      at Context.then (http://localhost:8081/__cypress/runner/cypress_runner.js:54262:21)
      at Context.<anonymous> (http://localhost:8081/__cypress/runner/cypress_runner.js:61169:21)
      at http://localhost:8081/__cypress/runner/cypress_runner.js:60874:33
      at tryCatcher (http://localhost:8081/__cypress/runner/cypress_runner.js:127195:23)
      at Promise._settlePromiseFromHandler (http://localhost:8081/__cypress/runner/cypress_runner.js:125213:31)
      at Promise._settlePromise (http://localhost:8081/__cypress/runner/cypress_runner.js:125270:18)
      at Promise._settlePromiseCtx (http://localhost:8081/__cypress/runner/cypress_runner.js:125307:10)
      at Async._drainQueue (http://localhost:8081/__cypress/runner/cypress_runner.js:122124:12)
      at Async._drainQueues (http://localhost:8081/__cypress/runner/cypress_runner.js:122129:10)
      at Async.drainQueues (http://localhost:8081/__cypress/runner/cypress_runner.js:122003:14)
      at <anonymous>
  (Results)
  ┌──────────────────���────────────────────┐
  │ Tests:        8                       │
  │ Passing:      7                       │
  │ Failing:      1                       │
  │ Pending:      0                       │
  │ Skipped:      0                       │
  │ Screenshots:  0                       │
  │ Video:        false                   │
  │ Duration:     1 minute, 3 seconds     │
  │ Spec Ran:     dashboard/index.test.js │
  └───────────────────────────────────────┘
================================================================================
  (Run Finished)
      Spec                                    Tests  Pass…  Fail…  Pend…  Skip… 
  ┌────────────────────────────────────────────────────────────────────────────┐
  │ ✖ dashboard/index.test.js         01:03      8      7      1      -      - │
  └────────────────────────────────────────────────────────────────────────────┘
    1 of 1 failed (100%)              01:03      8      7      1      -      -  
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! superset@0.999.0-dev cypress: `cypress "run" "--spec" "cypress/integration/dashboard/*" "--record" "false" "--config" "video=false"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the superset@0.999.0-dev cypress script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/travis/.npm/_logs/2019-04-23T02_11_51_043Z-debug.log
ERROR: InvocationError for command /home/travis/build/apache/incubator-superset/superset/assets/cypress_build.sh dashboard (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   cypress-dashboard: commands failed
```

cgivre commented:
@villebro @coderfender Thank you both for your help.  All unit tests passed!

cgivre commented:
@mistercrunch, I believe all the issues you've raised were addressed.  Could you please take a look?  Thanks!

coderfender commented:
I have a small question though (out of curiosity). Do we need to write Unit tests on the Python side to check this behavior ?

villebro commented on superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx:
```diff
@@ -83,17 +83,9 @@ export default class SqlEditorLeftBar extends React.PureComponent {
-    }
+    const tableName = tableOpt.value;
+    const schemaName = this.props.queryEditor.schema;
+    this.setState({ tableName });
```
A small comment here @mistercrunch : It's my understanding that some databases pass on the schema name along with the table name separated by a period (I believe BigQuery does, and Presto sometimes, too). It was apparently assumed that a period *always* indicates `schema.table`. However, periods are supported in many dbs schema/table names, hence they can't be universally removed. I don't removing this will have adverse side-effects, other than showing the shema name along with the table name on some dbs.

villebro commented:
@cgivre can you also add Drill in the table of supported engines in `docs/installation.rst` and any additional steps needed to get it up and running (if needed).

cgivre committed:
Updated Docs

villebro commented on docs/installation.rst:
```diff
@@ -449,6 +455,31 @@ Required environment variables: ::
+
+For a complete tutorial about how to use Apache Drill with Superset, see this tutorial:
+`Visualize Anything with Superset and Drill <http://thedataist.com/visualize-anything-with-superset-and-drill/>`_
+
```
Very thorough, nice!

cgivre replied:
Thank you!

cgivre commented on superset/assets/spec/javascripts/components/TableSelector_spec.jsx:
```diff
@@ -199,7 +199,7 @@ describe('TableSelector', () => {
         label: 'my_table',
       });
-      expect(mockedProps.onTableChange.getCall(0).args[0]).toBe('my_table');
+      expect(mockedProps.onTableChange.getCall(0).args[0]).toBe('main.my_table');
```
Done

cgivre commented:
@villebro @mistercrunch @kristw 
I believe all comments have been addressed.  Could someone please take a look and let me know if anything else needs to be addressed?

coderfender commented:
I can confirm that I have tested this on my cluster using S3 and File system (DFS and CP) plugins. Apart from issues in the Drill Dialect ( which isn't pertinent to this repo/project), I can confirm that the functionality seems to be as expected .

mistercrunch commented on superset/assets/src/components/TableSelector.jsx:
```diff
@@ -170,13 +170,8 @@ export default class TableSelector extends React.PureComponent {
-      tableName = namePieces[1];
-    }
+    const tableName = tableOpt.value;
+    const schemaName = this.props.schema;
```
I think if/when using `TABLE_NAMES_CACHE_CONFIG`, the logic that got removed will be needed.

villebro replied:
I tested this more thoroughly; it seems table name caching isn't affected, however BigQuery table preview did break due to there being a query of the form ``SELECT ... FROM `schema`.`schema.table` ``. OTOH e.g. the Postgres inspector doesn't append schema to table name. This got me thinking; instead of handling removal of schema from table name in the frontend, shouldn't this be done in the backend? I added an override to the BigQuery spec based on `BaseEngineSpec.get_table_names` method  to remove the schema name after inspection, and it worked perfectly (building on the PR branch):

```python
    @classmethod
    def get_table_names(cls, inspector, schema):
        tables = inspector.get_table_names(schema)
        return sorted([table.split('.')[1] for table in tables])
``` 
This has the added benefit of removing the unnecessary schema name from the table dropdown, and would make it possible to customize this behavior per engine.

cgivre replied:
@villebro Thanks for your help.  I would agree that the table handling/splitting should happen on the backend, so that way custom parsing (such as what I had to do for Drill) could all happen on in the backend.
@mistercrunch, does this complete the PR?

villebro commented:
@cgivre as this PR in its current state would break support for existing engines I don't think it can be merged. I would propose putting through a PR first that adds support for period separated table names, and once that's merged, rebasing this and merging once the dust settles. I have an idea of how this could be done with minimal collateral damage. @mistercrunch if this sounds ok I can put through a PR that makes minimal changes required to add support for period separated table names.

mistercrunch commented:
I think the feature around `TABLE_NAMES_CACHE_CONFIG` when enabled makes it such that on the left panel, when schema is NOT selected, the table dropdown, instead of behind greyed out, will allow to search across all the tables in the database across schemas. This concatenates the `{schema}.{table_name}`.

Now if we start having table names with periods in them, this stuff will get messed up. We'll either need proper quoting and quoting-aware splitting, or we need a better data structure, from say:

```json
[
    "myschema.table1",
    "myschema2.table2"
]
```
to:
```json
    {"schema": "myschema", "table": "table1"},
    {"schema": "myschema2", "table": "table2"}
```

villebro commented:
Thanks for the comment @mistercrunch , I had totally missed the `allow_multi_schema_metadata_fetch` db feature. Agree that this is non-trivial and needs to be thought through thoroughly.

cgivre commented:
I would be an advocate of the second approach that @mistercrunch mentioned where we have a better data structure.  However, I'm not sure I know enough about the internals of Superset to implement those kinds of changes.  (Unless I'm wrong and that would be an easy fix)

It also does seem out of scope for this PR.  Should we create a separate PR for that specific issue? @villebro could you assist with the implementation?

villebro commented:
@cgivre yes I do think this is out of scope for this PR. I can look into how this could be done. I also think the second approach is better.

cgivre commented:
Thanks @villebro!  I appreciate your help on this.

villebro commented on superset/db_engine_specs.py:
```diff
@@ -721,6 +721,70 @@ def get_table_names(cls, inspector, schema):
+        if '/' in uri.database:
+            database = uri.database.split('/')[0]
+        if selected_schema:
+            uri.database = database + '/' + selected_schema
```
As of #7297 is is recommended to apply escaping to `selected_schema` before adding it to the database uri. Take a look at how other connectors have done it, but something like this before changing `uri.database`:
```python
selected_schema = parse.quote(selected_schema, safe='')
```

cgivre replied:
Done

villebro commented on superset/db_engine_specs.py:
```diff
@@ -721,6 +721,70 @@ def get_table_names(cls, inspector, schema):
+
+    @classmethod
+    def extra_table_metadata(cls, database, table_name, schema_name):
+        return super().extra_table_metadata(database, table_name, schema_name)
```
If this merely calls `super()`, it can probably be omitted.

cgivre replied:
Done

villebro commented on superset/db_engine_specs.py:
```diff
@@ -721,6 +721,70 @@ def get_table_names(cls, inspector, schema):
+
+        return super().select_star(my_db, table_name, engine,
+                                   schema, limit,
+                                   show_cols, indent, latest_partition, cols)
```
same here

cgivre replied:
Fixed... I actually removed this method.

villebro commented on superset/db_engine_specs.py:
```diff
@@ -721,6 +721,70 @@ def get_table_names(cls, inspector, schema):
+
+    @classmethod
+    def get_table_names(cls, inspector, schema):
+        return sorted(inspector.get_table_names(schema))
```
this should also be removed, as #7453 changes the base method, and this is merely a copy of the original method.

villebro commented on superset/db_engine_specs.py:
```diff
@@ -721,6 +721,70 @@ def get_table_names(cls, inspector, schema):
+
+    @classmethod
+    def get_view_names(cls, inspector, schema):
+        return sorted(inspector.get_view_names(schema))
```
..and same here.

villebro commented on superset/db_engine_specs.py:
```diff
@@ -721,6 +721,70 @@ def get_table_names(cls, inspector, schema):
 
+class DrillEngineSpec(BaseEngineSpec):
+    """Engine spec for Apache Drill"""
+    engine = 'drill'
```
If you get around to trying out #7453 , try adding the following here:
```python
try_remove_schema_from_table_name = False
```
This will disable all attempts to remove schema from table names. While the risk of this happening are small (the table would have to be prefixed with the schema name followed by a period), it's better to disable schema name removal alltogether for engines that never add the schema name to table names.

cgivre replied:
@villebro 
I'm going to wait until #7453 is committed.  My experiments with your branch basically solved all the issues with the Drill integration.  Once that's done, this PR will just be some mods to the `db_engine_specs.py` and hopefully will be easy to commit as well.

cgivre replied:
I think I'm going to leave this out for the moment as Drill allows some schema variants that it might break.

villebro reviewed (changes requested):
I left some comments to help you test this PR with #7453 and clean up unnecessary changes. If and when #7453 gets merged, this PR should only have to modify `docs/installation.rst` and `superset/db_engine_specs.py`, as the javascript parts are already tackled by #7453.

cgivre committed:
Removed Extraneous Functions

cgivre committed:
Removed Extraneous Functions

cgivre committed:
Final Mods

cgivre reviewed (commented):
I made all the requested changes.   At this point, the PR really is only overwriting the functions relating the time-grain functionality for Drill.  Everything seems to work without any mods outside of the `db_engine_specs.py` file.

cgivre committed:
Fixed Unit Test Error

villebro commented on superset/db_engine_specs.py:
```diff
@@ -724,6 +724,46 @@ def get_table_names(cls, inspector, schema):
+    # Returns a function to convert a Unix timestamp in milliseconds to a date
+    @classmethod
+    def epoch_to_dttm(cls):
+        return 'TO_DATE({col})'
```
`epoch_to_dttm` should convert epoch in seconds to dttm, not milliseconds (the method `epoch_ms_to_dttm` is reserved for that). I'm thinking something along the following lines could be the correct solution?
```python
    @classmethod
    def epoch_to_dttm(cls):
        return cls.epoch_ms_to_dttm().replace('{col}', '({col}*1000)')

    @classmethod
    def epoch_ms_to_dttm(cls):
        return 'TO_DATE({col})'
```

cgivre replied:
Fixed.

cgivre committed:
Fixed Epoch Conversion Functions

villebro reviewed (approved):
LGTM, when you feel this is complete please remove WIP from the title.

cgivre commented:
@villebro I updated the title.  Thank you very much for your assistance getting this PR completed.  From my perspective we are ready to go.

villebro commented:
Let's leave this open for a few days to see if anyone has something to add. But IMO this is ready to be merged.

mistercrunch reviewed (approved):
LGTM!

SamuelMarks commented:
Awesome to see. Should the [README.md#database-support](https://github.com/apache/incubator-superset/blob/e4f8444/README.md#database-support) be updated with [Apache Drill](https://drill.apache.org)?

mistercrunch commented:
@cgivre ^^^

cgivre commented:
Thanks @mistercrunch and @SamuelMarks 
I addressed this and one small bug in https://github.com/apache/incubator-superset/pull/7635, which @villebro just approved.
